src/_pytest/python.py

```diff
@@ -1752,12 +1753,31 @@ def function(self):
         """Underlying python 'function' object."""
         return getimfunc(self.obj)
 
-    def _getobj(self):
-        assert self.parent is not None
+    @property
+    def instance(self):
+        try:
+            return self._instance
+        except AttributeError:
+            if isinstance(self.parent, Class):
+                # Each Function gets a fresh class instance.
+                self._instance = self._getinstance()
+            else:
+                self._instance = None
+        return self._instance
+
+    def _getinstance(self):
```

Comment:
I suggest we rename this method to make it clear it might create a new instance when called, perhaps `_new_instance`.

Reply:
The name `_getinstance` is meant to evoke `_getobj`. Since it's modeled after it I chose to make it similar.